yjit/src/codegen.rs

```diff
@@ -6227,6 +6135,155 @@ fn gen_send_iseq(
+    )
+}
+
+#[must_use]
+fn exit_if_supplying_kw_and_has_no_kw(asm: &mut Assembler, supplying_kws: bool, iseq: *const rb_iseq_t) -> Option<()> {
+    // If we have keyword arguments being passed to a callee that only takes
+    // positionals, then we need to allocate a hash. For now we're going to
+    // call that too complex and bail.
+    exit_if(
+        asm,
+        supplying_kws && !unsafe { get_iseq_flags_has_kw(iseq) },
+        Counter::send_iseq_has_no_kw,
+    )
+}
+
+#[must_use]
+fn exit_if_supplying_kws_and_accept_no_kwargs(asm: &mut Assembler, supplying_kws: bool, iseq: *const rb_iseq_t) -> Option<()> {
+    // If we have a method accepting no kwargs (**nil), exit if we have passed
+    // it any kwargs.
+    exit_if(
+        asm,
+        supplying_kws && unsafe { get_iseq_flags_accepts_no_kwarg(iseq) },
+        Counter::send_iseq_accepts_no_kwarg
+    )
+}
+
+#[must_use]
+fn exit_if_splat_and_zsuper(asm: &mut Assembler, flags: u32) -> Option<()> {
+    // zsuper methods are super calls without any arguments.
+    // They are also marked as splat, but don't actually have an array
+    // they pull arguments from, instead we need to change to call
+    // a different method with the current stack.
+    exit_if(asm, flags & VM_CALL_ARGS_SPLAT != 0 && flags & VM_CALL_ZSUPER != 0, Counter::send_iseq_zsuper)
+}
+
+#[must_use]
+fn exit_if_doing_kw_and_splat(asm: &mut Assembler, doing_kw_call: bool, flags: u32) -> Option<()> {
+    exit_if(asm, doing_kw_call && flags & VM_CALL_ARGS_SPLAT != 0, Counter::send_iseq_splat_with_kw)
+}
+
+#[must_use]
+fn exit_if_wrong_number_arguments(asm: &mut Assembler, opts_filled: i32, flags: u32, opt_num: i32, iseq_has_rest: bool) -> Option<()> {
+    // Too few arguments and no splat to make up for it
+    let too_few = opts_filled < 0 && flags & VM_CALL_ARGS_SPLAT == 0;
+    // Too many arguments and no place to put them (i.e. rest arg)
+    let too_many = opts_filled > opt_num && !iseq_has_rest;
+
+    exit_if(asm, too_few || too_many, Counter::send_iseq_arity_error)
+}
+
+#[must_use]
+fn exit_if_doing_kw_and_opts_missing(asm: &mut Assembler, doing_kw_call: bool, opts_missing: i32) -> Option<()> {
+    // If we have unfilled optional arguments and keyword arguments then we
+    // would need to adjust the arguments location to account for that.
+    // For now we aren't handling this case.
+    exit_if(asm, doing_kw_call && opts_missing > 0, Counter::send_iseq_missing_optional_kw)
+}
+
+#[must_use]
+fn exit_if_has_rest_and_optional_and_block(asm: &mut Assembler, iseq_has_rest: bool, opt_num: i32, iseq: *const rb_iseq_t, block_arg: bool) -> Option<()> {
+    exit_if(
+        asm,
+        iseq_has_rest && opt_num != 0 && (unsafe { get_iseq_flags_has_block(iseq) } || block_arg),
+        Counter::send_iseq_has_rest_opt_and_block
+    )
+}
+
+#[must_use]
+fn exit_if_unsupported_block_arg_type(asm: &mut Assembler, block_arg_type: Option<Type>) -> Option<()> {
```

Comment:
`gen_send_cfunc` could use this too, but is it out of scope in this PR?

Reply:
I assume cfunc and iseq will probably get support for block types at different times. So I didn't extract that out into a function.

Reply:
[This place](https://github.com/ruby/ruby/blob/165ef42b3c29414a29c3720bd353c990bb9a6638/yjit/src/codegen.rs#L5545-L5556) currently has the exact same code, so I thought you could just call `exit_if_unsupported_block_arg_type(asm, block_arg_type)?;` there, but do you mean it's gonna change soon?

Reply:
But adding support for other block types to cfunc and iseq at different times does make sense. If sharing it makes it more difficult, it's fine to leave it as is. I wonder how you're going to do the same kind of refactoring on `gen_send_cfunc` though (the function names will conflict if it's not shared).